--- samba-3.0.13/source/client/client.c.smbclient	2005-04-05 13:51:47.000000000 -0400
+++ samba-3.0.13/source/client/client.c	2005-04-05 13:52:46.000000000 -0400
@@ -3177,6 +3177,7 @@
 		   else but port 139... */
 
 		cli_cm_shutdown();
+		port = 139;
 		cli_cm_set_port( 139 );
 		cli = cli_cm_open(query_host, "IPC$", True);
 	}
@@ -3339,8 +3340,9 @@
 			name_type = 0x03; 
 			cli_cm_set_dest_name_type( name_type );
 			pstrcpy(desthost,poptGetOptArg(pc));
-			if( !port )
-				cli_cm_set_port( 139 );
+			if ( port != 139 )
+				port = 139;
+			cli_cm_set_port ( 138);
  			message = True;
  			break;
 		case 'I':
