diff --git a/source/nsswitch/idmap_ldap.c b/source/nsswitch/idmap_ldap.c
index ca7d32b..7c02779 100644
--- a/source/nsswitch/idmap_ldap.c
+++ b/source/nsswitch/idmap_ldap.c
@@ -287,16 +287,9 @@ static NTSTATUS idmap_ldap_alloc_init(const char *params)
 	}
 	CHECK_ALLOC_DONE( idmap_alloc_ldap->url );
 
-	tmp = lp_ldap_idmap_suffix();
+	tmp = lp_parm_const_string(-1, "idmap alloc config", "ldap_base_dn", NULL);
 	if ( ! tmp || ! *tmp) {
-		tmp = lp_parm_const_string(-1, "idmap alloc config", "ldap_base_dn", NULL);
-	}
-	if ( ! tmp) {
-		tmp = lp_ldap_suffix();
-		if (tmp) {
-			DEBUG(1, ("WARNING: Trying to use the global ldap suffix(%s)\n", tmp));
-			DEBUGADD(1, ("as suffix. This may not be what you want!\n"));
-		}
+		tmp = lp_ldap_idmap_suffix();
 		if ( ! tmp) {
 			DEBUG(1, ("ERROR: missing idmap ldap suffix\n"));
 			ret = NT_STATUS_UNSUCCESSFUL;
@@ -780,21 +773,16 @@ static NTSTATUS idmap_ldap_db_init(struct idmap_domain *dom)
 	}
 	CHECK_ALLOC_DONE(ctx->url);
 
-	tmp = lp_ldap_idmap_suffix();
-	if ( ! tmp || ! *tmp) {
-		tmp = lp_parm_const_string(-1, config_option, "ldap_base_dn", NULL);
-	}
-	if ( ! tmp) {
-		tmp = lp_ldap_suffix();
-		if (tmp) {
-			DEBUG(1, ("WARNING: Trying to use the global ldap suffix(%s)\n", tmp));
-			DEBUGADD(1, ("as suffix. This may not be what you want!\n"));
-		} else {
-			DEBUG(1, ("ERROR: missing idmap ldap suffix\n"));
-			ret = NT_STATUS_UNSUCCESSFUL;
-			goto done;
-		}		
-	}
+        tmp = lp_parm_const_string(-1, config_option, "ldap_base_dn", NULL);
+        if ( ! tmp || ! *tmp) {
+                tmp = lp_ldap_idmap_suffix();
+                if ( ! tmp) {
+                        DEBUG(1, ("ERROR: missing idmap ldap suffix\n"));
+                        ret = NT_STATUS_UNSUCCESSFUL;
+                        goto done;
+		}
+        }
+
 	ctx->suffix = talloc_strdup(ctx, tmp);
 	CHECK_ALLOC_DONE(ctx->suffix);
 
