From 26183cf4904d76962b218401fe07a9f7642e05ff Mon Sep 17 00:00:00 2001
From: Andreas Schneider <asn@samba.org>
Date: Wed, 5 Jul 2017 08:59:23 +0200
Subject: [PATCH] unittests: Add missing stdint.h include

Signed-off-by: Andreas Schneider <asn@samba.org>
---
 testsuite/unittests/test_sambafs_srv_pipe.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/testsuite/unittests/test_sambafs_srv_pipe.c b/testsuite/unittests/test_sambafs_srv_pipe.c
index 641e99d..553f530 100644
--- a/testsuite/unittests/test_sambafs_srv_pipe.c
+++ b/testsuite/unittests/test_sambafs_srv_pipe.c
@@ -2,6 +2,7 @@
 #include <stdarg.h>
 #include <stddef.h>
 #include <setjmp.h>
+#include <stdint.h>
 #include <unistd.h>
 #include <sys/stat.h>
 #include <cmocka.h>
-- 
2.9.4

