--- source/winbindd/winbindd.c	2007-09-24 00:01:39 +0000
+++ source/winbindd/winbindd.c	2007-10-17 07:58:37 +0000
@@ -532,8 +532,8 @@
 	}
 
 	if (*(uint32 *)(&state->request) != sizeof(state->request)) {
-		DEBUG(0,("request_len_recv: Invalid request size received: %d\n",
-			 *(uint32 *)(&state->request)));
+		DEBUG(0,("request_len_recv: Invalid request size received: %d (expected %d)\n",
+			 *(uint32 *)(&state->request), sizeof(state->request)));
 		state->finished = True;
 		return;
 	}

--- source/nsswitch/winbind_struct_protocol.h	2007-04-20 23:12:49 +0000
+++ source/nsswitch/winbind_struct_protocol.h	2007-10-17 07:58:37 +0000
@@ -326,7 +326,7 @@
 		   The size is the sizeof the union without the padding aligned on 
 		   an 8 byte boundary.   --jerry */
 
-		char padding[1560];
+		char padding[1800];
 	} data;
 	union {
 		SMB_TIME_T padding;

